Sources/AsyncAlgorithms/AsyncShareSequence.swift

```diff
@@ -0,0 +1,693 @@
+      var iteratingTask: IteratingTask
+      private(set) var buffer = Deque<Element>()
+      private(set) var finished = false
+      private(set) var failure: Failure?
+      var cancelled = false
+      var limit: UnsafeContinuation<Bool, Never>?
+      var demand: UnsafeContinuation<Void, Never>?
+      
+      let storagePolicy: StoragePolicy
+      
+      init(_ iteratorFactory: @escaping @Sendable () -> sending Base.AsyncIterator, bufferingPolicy: AsyncBufferSequencePolicy) {
+        self.iteratingTask = .pending(iteratorFactory)
+        switch bufferingPolicy.policy {
+        case .bounded: self.storagePolicy = .unbounded
+        case .bufferingOldest(let bound): self.storagePolicy = .bufferingOldest(bound)
+        case .bufferingNewest(let bound): self.storagePolicy = .bufferingNewest(bound)
+        case .unbounded: self.storagePolicy = .unbounded
+        }
+      }
+      
+      // Removes elements from the front of the buffer that all consumers have already processed.
+      //
+      // This method finds the minimum position across all active consumers and removes
+      // that many elements from the front of the buffer. It then adjusts all consumer
+      // positions to account for the removed elements, maintaining their relative positions.
+      //
+      // This optimization prevents the buffer from growing indefinitely when all consumers
+      // are keeping pace with each other.
+      mutating func trimBuffer() {
+        if let minimumIndex = sides.values.map({ $0.position }).min(), minimumIndex > 0 {
+          buffer.removeFirst(minimumIndex)
+          sides = sides.mapValues {
+            $0.offset(minimumIndex)
+          }
+        }
+      }
+      
+      // Private state machine transitions for the emission of a given value.
+      //
+      // This method ensures the continuations are properly consumed when emitting values
+      // and returns those continuations for resumption.
+      private mutating func _emit<T>(_ value: T, limit: Int) -> (T, UnsafeContinuation<Bool, Never>?, UnsafeContinuation<Void, Never>?, Bool) {
+        let belowLimit = buffer.count < limit || limit == 0
+        defer {
+          if belowLimit {
+            self.limit = nil
+          }
+          demand = nil
+        }
+        if case .cancelled = iteratingTask {
+          return (value, belowLimit ? self.limit : nil, demand, true)
+        } else {
+          return (value, belowLimit ? self.limit : nil, demand, false)
+        }
+      }
+      
+      // Internal state machine transitions for the emission of a given value.
+      //
+      // This method ensures the continuations are properly consumed when emitting values
+      // and returns those continuations for resumption.
+      //
+      // If no limit is specified it interprets that as an unbounded limit.
+      mutating func emit<T>(_ value: T, limit: Int?) -> (T, UnsafeContinuation<Bool, Never>?, UnsafeContinuation<Void, Never>?, Bool) {
```

Comment:
I find it really confusing that this method is called `emit` since it is called from the consumer from what I can tell. Can we call this `next` or something with `next` instead`?